### ISSUE TITLE
FQF logit computation

### ISSUE THREAD

**Issue by dmartinezbaselga** [opendilab/DI-engine#775]
Hello!

Thank you for this project, it's really complete and modular, which makes it easy to replicate and modify the code. I have a comment regarding the implementation of FQF. As far as I understand the method, line 772 of the file DI-engine/ding/model/common/head.py should be something like:
```python
q_quantiles_width = q_quantiles[..., 1:] - q_quantiles[..., :-1]
logit = (q.permute(2,0,1)*q_quantiles_width).sum(2).permute(1,0)
```

instead of:
`logit = q.mean(1)`

Thank you in advance for your time, and sorry if I am wrong with the issue and it's a misunderstanding!

**Comment by PaParaZz1** [comment]
I think we have already implemented the operation you mentioned line 763-768 ([link](https://github.com/opendilab/DI-engine/blob/main/ding/model/common/head.py#L763)), i.e.:
```python
        q_quantiles_hats = (q_quantiles[:, 1:] + q_quantiles[:, :-1]).detach() / 2.  # (batch_size, num_quantiles)

        # NOTE(rjy): reparameterize q_quantiles_hats
        q_quantile_net = self.quantile_net(q_quantiles_hats)  # [batch_size, num_quantiles, hidden_size(64)]
        # x.view[batch_size, 1, hidden_size(64)]
        q_x = (x.view(batch_size, 1, -1) * q_quantile_net)  # [batch_size, num_quantiles, hidden_size(64)]
```
Could you please confirm again? If you have other problems, you can continue to reply in this issue.

**Comment by dmartinezbaselga** [comment]
Hi,

Thanks for the response. What you are referring is the computation of $\hat{\tau_i}=\frac{\tau_{i+1}-\tau{i}}{2}$ (line 763) and the embedding computation of $\hat{\tau}$ (764-768), which is in the paper in section 3.4:
![image](https://github.com/opendilab/DI-engine/assets/44682605/78a47258-caeb-426b-90e0-16b30fd15666)
These refer to the quantile values computation:
![image](https://github.com/opendilab/DI-engine/assets/44682605/bdf15f7e-fb81-466d-b3f4-aacf6e57210b)
The part that I am missing is the $\tau_{i+1}-\tau_{i}$ that measures the width of the quantile fractions of:
![image](https://github.com/opendilab/DI-engine/assets/44682605/0247b4c3-708a-4ecf-a5c7-33872f855c0f)
Instead of a mean, it's a weighted mean.

**Comment by PaParaZz1** [comment]
Thanks for your feedback, I have checked this part of the original paper and our implementation.
- On the one hand, the operation you mentioned is implemented in line 773 ([link](https://github.com/opendilab/DI-engine/blob/main/ding/model/common/head.py#L773)). This variable is only used for the computation of target value.
- On the other hand, the variable `logit` is only used to select action in collect/eval mode, it is not used in training (learn mode).
